File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-6.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.ans: TreeNode = None
        self.traversed: int = 0
        def inorder(curr: TreeNode) -> None:
            if curr is None:
                return
            inorder(curr.left)
            self.traversed += 1
            if self.traversed == k:
                self.ans = curr
                return
            inorder(curr.right)
        inorder(root)
        return self.ans.val
        

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int curr_idx = 0;
    // unordered_set<int> seen;
    unordered_map<int, int> pos;
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int n = (int)inorder.size();
        for (int i = 0; i < n; i++) {
            pos[inorder[i]] = i;
        }
        TreeNode* root = buildNode(0, n - 1, preorder, inorder);
        return root;
    }
    TreeNode* buildNode(
        int left, int right, 
        vector<int>& preorder, vector<int>& inorder
    ) {
        if (left > right) {
            return nullptr;
        }
        TreeNode* root = new TreeNode(preorder[curr_idx]);
        int inorder_pos = pos[preorder[curr_idx]];
        curr_idx++;
        root->left = buildNode(left, inorder_pos - 1, preorder, inorder);
        root->right = buildNode(inorder_pos + 1, right, preorder, inorder);
        return root;
    }
};
